### PR TITLE
[MLDM-60] Implement content hashing for PJS

### PR DESCRIPTION
**Pull request by robert-uhl**
Hash filesets and filesystems.  Not yet hooked into job creation.

### CODE REVIEW

File: src/internal/client/storage.go [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+	"github.com/pachyderm/pachyderm/v2/src/internal/errors"
+)
+
+// FileSystemToFileset creates a new fileset containing fileSystem, returning the fileset ID.
+func (c APIClient) FileSystemToFileset(ctx context.Context, fileSystem fs.FS) (string, error) {
+	cc, err := c.FilesetClient.CreateFileset(ctx)
+	if err != nil {
+		return "", errors.Wrap(err, "could not create fileset")
+	}
+	if err := fs.WalkDir(fileSystem, ".", func(path string, d fs.DirEntry, err error) error {
```
Comment (FahadBSyed) [inline src/internal/client/storage.go:21]: Just to confirm, are entries returned lexicographically with directories?

So if the structure was:
```
/
a/
b
```
and the structure of a/ was
```
a/c
a/d
```
would we get:
```
a/c
a/d
b
```
in the output?

File: src/internal/pjs/hash.go [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+	return h.sum()
+}
+
+// A FilesetReader provides a store.Fileset_ReadFilesetClient for hashing.
+type FilesetReader interface {
+	ReadFileset(context.Context, *storage.ReadFilesetRequest, ...grpc.CallOption) (storage.Fileset_ReadFilesetClient, error)
+}
+
+// HashFileSetID hashes the fileset identified by ID from the FilesetReader fr.
+func HashFilesetID(ctx context.Context, fr FilesetReader, id string) ([]byte, error) {
```
Comment (FahadBSyed) [inline src/internal/pjs/hash.go:55]: [nit]: this sounds like we are hashing the ID itself. I don't think HashFilesetByID sounds much better, so I can't offer a better name.

File: src/internal/pjs/hash_test.go [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+	var testFS = fstest.MapFS{
+		"test":   &fstest.MapFile{Data: []byte("123")},
+		"abc123": &fstest.MapFile{Data: []byte("foo")},
+	}
+	id, err := env.PachClient.FileSystemToFileset(ctx, testFS)
+	require.NoError(t, err, "FileSystemToFileset")
+	h, err := pjs.HashFilesetID(ctx, env.PachClient.FilesetClient, id)
+	require.NoError(t, err, "hashing fileset")
+	require.Equal(t, "25ec7afdab18e8e5808730def751575fdb964d293a36ce21103415c747ca06db", hex.EncodeToString(h))
+}
```
Comment (FahadBSyed) [inline src/internal/pjs/hash_test.go:37]: Would be nice to see directory-like entries tested here, both in the mapFS and the fileset.

**Review by FahadBSyed — APPROVED** [review]
approved with some suggestions

File: src/internal/client/storage.go (hunk 1 above)
Comment (robert-uhl) [reply src/internal/client/storage.go:21]: [Yup](https://cs.opensource.google/go/go/+/refs/tags/go1.23.0:src/testing/fstest/mapfs.go;l=103-105)

File: src/internal/pjs/hash.go (hunk 2 above)
Comment (robert-uhl) [reply src/internal/pjs/hash.go:55]: `HashFileset` sounds good.

File: src/internal/pjs/hash_test.go (hunk 3 above)
Comment (robert-uhl) [reply src/internal/pjs/hash_test.go:37]: True — and a final test to ensure dynamic filesets and filesystems hash properly.

**Pull request merged**